widget/task_preview.lua

```diff
@@ -0,0 +1,120 @@
+local gears = require("gears")
+local beautiful = require("beautiful")
+local dpi = beautiful.xresources.apply_dpi
+local cairo = require("lgi").cairo
+
+local function draw_widget(c, task_preview_box, screen_radius, widget_bg,
+                           widget_border_color, widget_border_width, margin)
+
+    local content = gears.surface(c.content)
+    local cr = cairo.Context(content)
+    local x, y, w, h = cr:clip_extents()
+    local img = cairo.ImageSurface.create(cairo.Format.ARGB32, w - x, h - y)
+    cr = cairo.Context(img)
+    cr:set_source_surface(content, 0, 0)
+    cr.operator = cairo.Operator.SOURCE
+    cr:paint()
+
+    task_preview_box:setup{
+        {
+            {
+                {
+                    {
+                        image = gears.surface.load(c.icon),
+                        resize = true,
+                        forced_height = dpi(20),
+                        forced_width = dpi(20),
+                        widget = wibox.widget.imagebox
+                    },
+                    {
+                        markup = c.name,
+                        align = "center",
+                        widget = wibox.widget.textbox
+                    },
+                    layout = wibox.layout.align.horizontal
+                },
+                {
+                    {
+                        {
+                            image = gears.surface.load(img),
+                            resize = true,
+                            widget = wibox.widget.imagebox
+                        },
+                        valign = "center",
+                        halign = "center",
+                        widget = wibox.container.place
+                    },
+                    top = margin * 0.25,
+                    widget = wibox.container.margin
+                },
+                fill_space = true,
+                layout = wibox.layout.fixed.vertical
+            },
+            margins = margin,
+            widget = wibox.container.margin
+        },
+        bg = widget_bg,
+        border_width = widget_border_width,
+        border_color = widget_border_color,
+        shape = helpers.shape.rrect(screen_radius),
+        widget = wibox.container.background
+    }
+end
+
+local enable = function(opts)
+    local widget_x = dpi(20)
+    local widget_y = dpi(20)
+    local widget_height = dpi(200)
+    local widget_width = dpi(200)
+    local margin = beautiful.task_preview_widget_margin or dpi(0)
+    local screen_radius = beautiful.task_preview_widget_border_radius or dpi(0)
+    local widget_bg = beautiful.task_preview_widget_bg or "#000000"
+    local widget_border_color = beautiful.task_preview_widget_border_color or
+                                    "#ffffff"
+    local widget_border_width = beautiful.task_preview_widget_border_width or
+                                    dpi(3)
+    local placement_fn = nil
+
+    if opts then
```

Comment:
I think you could safe that block here if you do `local opts = opts or {}` at the beginning of the function and then put the defaults in the same line as possible user configs ie `local widget_x = opts.x or dpi(20)`. Not a big deal though